Add currency to support level

Sjors opened the pull request:
Fixes #3

Tested EUR with BTCPay server. I didn't test Stripe yet.

To make it work with BTC as a currency we'd also need to allow fractional values. It would also require a workaround for Stripe. However I don't want to make this PR more complicated.

I should probably add some tests.

JeffVandrewJr commented:
Localization is much appreciated! 

Please test Stripe (instructions for Stripe test charges: https://github.com/JeffVandrewJr/patron/blob/master/test-charge.md). You had also mentioned adding some tests; that is much appreciated as functiona;/unit tests are somewhat lacking right now.

If Stripe/other testing is successful on your end, please post here and I will then test everything in Docker to make sure none of these changes broke the primary Docker deployment. Assuming that all works, I will then marge.

JeffVandrewJr commented:
Update on the above comment: Square does not support multi-currency (https://squareup.com/help/us/en/article/5415-can-i-accept-multiple-currencies-with-square). Each account can only accept the home currency. Currently Square is available in USA, Canada, Japan, Australia, and UK. It is not available in the Eurozone.

Adding Euro/multi-currency support is still a good idea for BTC payments, however.

With that in mind, I would suggest adding a warning that using currencies other than USD will deactivate Square, and automatically make sure Square is inactive if a non-USD currency is selected. That would seem to be the easiest way to avoid breaking anything at the current time.

If you can add that, as well as any other tests you mentioned in your above comment, please then post here and I will test the main Docker deployment to ensure nothing is broken. Assuming not, I will then merge.

Thanks!

Sjors commented:
Made a minor tweak: automatically convert currency name to symbol, so it looks like this:

<img width="343" alt="Schermafbeelding 2019-05-06 om 18 04 30" src="https://user-images.githubusercontent.com/10217/57238326-69a67f00-7029-11e9-8ff2-a9d191220224.png">

Same on the creditcard page (see below).

This uses the Babel dependency which I added to `dependencies.txt`.

I also added that for Square the currency has to match:

<img width="844" alt="Schermafbeelding 2019-05-06 om 18 05 42" src="https://user-images.githubusercontent.com/10217/57238389-9490d300-7029-11e9-9e3f-31cbd0645013.png">

I'm not having much luck with Square, perhaps because I'm in the Eurozone. I setup a test shop and tried both a EUR and USD payment with the test credentials. Both failed.

<img width="402" alt="Schermafbeelding 2019-05-06 om 18 23 02" src="https://user-images.githubusercontent.com/10217/57239781-2d751d80-702d-11e9-9058-44734728ade7.png">

<img width="403" alt="Schermafbeelding 2019-05-06 om 18 29 23" src="https://user-images.githubusercontent.com/10217/57239793-3239d180-702d-11e9-875a-ffd312f2a3e8.png">

<img width="1329" alt="Schermafbeelding 2019-05-06 om 18 31 53" src="https://user-images.githubusercontent.com/10217/57239807-3cf46680-702d-11e9-9765-822c5010088d.png">

JeffVandrewJr commented:
@Sjors Yeah Square isn't going to work for Eurozone payments. Is there any way you can add a warning to customers that choose non-USD currencies that Square won't work? Also please undo the changes you made to the code that processes Square payments.

Sjors commented:
> Yeah Square isn't going to work for Eurozone payments.

Not on production, but I'm not sure if that explains the error in their developer sandbox. Someone else should test that.

>  Is there any way you can add a warning to customers that choose non-USD currencies that Square won't work?

It follows from the warning I added "The currency must match your Square home currency". We could also maybe fetch the currency via Square API as an extra check.

> please undo the changes you made to the code that processes Square payments.

The changes in `process_square` are still needed for GBP, AUD, etc.

JeffVandrewJr commented:
@Sjors I think adding the currency code to the `PriceLevel` model is as problem because price level data is set via freeform in the admin panel, so there is ample opportunity for the user to accidentally enter a bad three character currency code on a given price level which presumably will cause chaos on the backend.

One easy way to rectify this would be to have currency instead be an attribute of the BTCPay model, and therefore applied globally throughout the entire instance. It could be set by creating a new "Home Currency" view in `patron/app/admin_views/__init__.py` which would correspond display a form set in `patron/app/admin_views/forms.py`, said form being a dropdown of valid BTCPay currency codes. As a dropbdown rather than freeform text, this would force a valid currency code.

Sjors commented:
I agree a dropdown would be better than a freeform field. Does the Python framework you use support enums perhaps (that automatically generate a dropdown form)?

JeffVandrewJr commented:
@Sjors Flask is unopinionated on this and basically all matters (unlike Django). LibrePatron uses Flask-WTForms to accomplish this; if you want to check out how to do it using WTForms check out a `SelectField` in the WTForms documentation. It's very easy; I used one in `app/admin_views/forms.py` in the `ThemeForm` if you want to see an example. 

`SelectField` basically uses tuples comprised of values paired with the display text for each value for the drop-down.
